src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,430 @@
+            _connectTcs = null;
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownBegin(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Shutdown begin {MsQuicConstants.ErrorTypeFromErrorCode(connectionEvent.ShutdownBeginStatus)}");;
+
+            _connectTcs?.SetResult(null);
+            _connectTcs = null;
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownBeginPeer(ConnectionEvent connectionEvent)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Shutdown Complete");
+
+            _shutdownTcs?.SetResult(null);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventNewStream(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            MsQuicStream msQuicStream = new MsQuicStream(_api, this, connectionEvent.StreamFlags, connectionEvent.Data.NewStream.Stream, inbound: true);
+
+            _acceptQueue.Writer.TryWrite(msQuicStream);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventStreamsAvailable(ConnectionEvent connectionEvent)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        internal override async ValueTask<QuicStreamProvider> AcceptStreamAsync(CancellationToken cancellationToken = default)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (await _acceptQueue.Reader.WaitToReadAsync(cancellationToken))
+            {
+                if (_acceptQueue.Reader.TryRead(out MsQuicStream stream))
+                {
+                    if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+                    return stream;
+                }
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return null;
+        }
+
+        internal override QuicStreamProvider OpenUnidirectionalStream()
+        {
+            return StreamOpen(QUIC_STREAM_OPEN_FLAG.UNIDIRECTIONAL);
+        }
+
+        internal override QuicStreamProvider OpenBidirectionalStream()
+        {
+            return StreamOpen(QUIC_STREAM_OPEN_FLAG.NONE);
+        }
+
+        public unsafe void SetIdleTimeout(TimeSpan timeout)
```

Comment:
Should all these be Properties (get and set) instead of functions?

Reply:
Maybe, I think I'm going to add options types to both QuicConnection and QuicListener. I don't think I want this function public and have the idle timeout as a parameter in the options.

Reply:
@geoffkizer cool if I make some options types?